### ISSUE TITLE
Add TypeScript typings

### ISSUE THREAD

**Issue by rkuykendall** [formsy/formsy-react#10]
Original issue: https://github.com/christianalfoni/formsy-react/issues/394

**Comment by mulmulovich** [comment]
any update about the typings?

**Comment by rkuykendall** [comment]
Hey @mulmulovich! Nobody is currently working on this. I actually have no idea how to add typings to a project, but would like to do this eventually. If someone can help with the basic structure, I could definitely help fill in all the correct signatures.

I've been using this interface in the little Typescript I'm writing now:

```ts
export default interface IFormsyDecorator {
  getErrorMessage (): any;
  getValue (): any;
  hasValue (): boolean;
  isFormDisabled (): boolean;
  isFormSubmitted (): boolean;
  isPristine (): boolean;
  isRequired (): boolean;
  isValid (): boolean;
  isValidValue (): boolean;
  resetValue (): void;
  setValidations (validations: any, required: boolean): void;
  setValue (value: any): void;
  showError (): boolean;
  showRequired (): boolean;
}
```

**Comment by Nibblesh** [comment]
@rkuykendall Although a lot of effort, writing the entire library in typescript would mean that the build process would output JS + definitions. Although this would be a fair chunk of work it would mean that further maintenance of the project would keep the Typescript and JS users happy with minimal effort.

**Comment by MilosRasic** [comment]
Writing the library in typescript would require contributors to also work with typescript.

I've seen and edited some seemingly solid typings in https://github.com/YouCanBookMe/react-datetime before.  They have both v1.8 in react-datetime.d.ts and v2 typings in DateTime.d.ts. They are covered by tests, at least v2. Don't have much experience with ts so I don't know how good this is, but it's definitely better than what we have for formsy. Either way, help is definitely wanted and needed.

**Comment by robbyemmert** [comment]
😭 This is a must for me.

**Comment by mohithg** [comment]
Typing is most appreciated for this project.

**Comment by rkuykendall** [comment]
Update: I've actually learned how to do typings while working on another project, so I'll be able to write them for Formsy once I have some free time.

**Comment by danzel** [comment]
Here are the types I've made, not sure they are totally correct but may be a further start

```ts
type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

declare module 'formsy-react' {

  export interface FormsyMixinProps<T> {
    value: T | undefined;
  }

  export interface FormsyInjectedProps<T> {
    getValue: () => T;
    setValue: (value: T | undefined) => void;
    getErrorMessage: () => string;
    showRequired: () => boolean;
    showError: () => boolean;
  }

  export function withFormsy<T, P extends FormsyInjectedProps<T>>(component: React.ComponentClass<P>): React.ComponentClass<Omit<P & FormsyMixinProps<T>, keyof FormsyInjectedProps<T>>>;

  export interface FormsyProps<TData> {
    onSubmit?: (data: TData, resetForm: boolean, invalidateForm: (propertyErrors: { [key: string]: string }) => void) => void;
    validationErrors?: { [key: string]: string };
    className?: string;
  }

  export default class Formsy<T> extends React.Component<FormsyProps<T>> {
  }
}
```

For components you wrap in `withFormsy` they should look like this:

```ts
export interface InputProps extends FormsyInjectedProps<string> {
  ...
}

class Input extends React.Component<InputProps> {
  ...
}

export default withFormsy<string, InputProps>(Input);
```

Then when you use the `<Input>` component you need to specify the `value` prop, but inside the `<Input>` component, you have the `getValue()` and `setValue()` props etc.

**Comment by rkuykendall** [comment]
@mulmulovich @Nibblesh @robbyemmert @mohithg @danzel 

I've finished the overhauls to the build and other systems to make this possible. I'm now in the process of converting the library to Typescript. Are any of you able to help me with this process?

I'm currently down to 38 build errors and will continue making progress today, but I'm using `any` pretty heavily and would appreciate anyone who wants to do a shift.

Alternatively, just anyone willing to code review my PR or use the 2.x version after merge would be very appreciated.

**Comment by rkuykendall** [comment]
Decided to do it piecemeal. First PR done: https://github.com/formsy/formsy-react/pull/174

**Comment by rkuykendall** [comment]
Closed this on accident. I'll close it only once we've shipped a v2 beta with full types which I'll post about here, but they may be weak at the beginning (lots of anys) and tighten up over time.

**Comment by rkuykendall** [comment]
Published the first beta release with types,  `formsy-react@2.0.0-beta.4`.

These types have not been tested yet ( I'm in the process of upgrading my own codebase now ) so there will very likely be issues, especially since typing a HOC is really, really complicated. Any help testing or improving the types would be very appreciated.

**Comment by rafaelhrsantos** [comment]
I've given a try to the new beta version but I've found some possible issues with the typings:

The interface WrapperProps has all fields set as required which is not true, at least I've been using the library with "name" as the only required field. So right now I've redefined the type to this on my side:
```
  interface WrapperProps {
    innerRef?: (ref: any) => void;
    name: string;
    required?: string | boolean;
    validationError?: any;
    validationErrors?: any;
    validations?: Validations | string;
    value?: any;
  }
```

I also had to do some adjustments on the withFormsy HOC since it won't accept FunctionComponents as it is right now. I've change to HOC to this:
```
type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
function withFormsy<T extends PassDownProps>(component: React.ElementType<T>):
    React.ComponentType<Omit<T, keyof PassDownProps> & WrapperProps>;
```

**Comment by rkuykendall** [comment]
@rafaelhrsantos How does this look? https://github.com/formsy/formsy-react/pull/259/files

**Comment by rkuykendall** [comment]
New types published thanks to @rafaelhrsantos ! Everyone here please try them out and comment! This is the only thing holding up 2.x going live!

https://github.com/formsy/formsy-react/releases/tag/v2.0.0-beta.5

**Comment by jlefebure** [comment]
On going from my side (upgrade from 1 to 2, typescript react project.
Anyway, why not passing value prop to a generic type ? That would avoid a lot of trouble

**Comment by jlefebure** [comment]
Okay, so everything looks great, just the addValidationRule type seems to be incorrect
From the Formsy docs (v1), the behaviour is to return either a boolean which indicate if the field is valid or not, or a string which specify an error message. 

```jsx
<MyInput name="email" validations="isEmail"/>
<MyInput name="number" validations="isNumeric,isLength:5"/>
<MyInput name="number" validations={{
  isNumeric: true,
  isLength: 5
}}/>
<MyInput name="number" validations={{
  myCustomIsFiveValidation: function (values, value) {
    values; // Other current values in form {foo: 'bar', 'number': 5}
    value; // 5
    return 5 === value ? true : 'No five'; // You can return an error
  }
}}/>
```

But the type of the ValidationFunction specify that the return type must be a boolean. So in TS, the previous code does not work anymore.

I would suggest to change the following type from 
```jsx
export declare type ValidationFunction = (values: Values, value: Value, extra?: any) => boolean;
```

to 
```jsx
export declare type ValidationFunction = (values: Values, value: Value, extra?: any) => boolean | string;
```

**Comment by rkuykendall** [comment]
@jlefebure Very open! It would be very much appreciated.

**Comment by rkuykendall** [comment]
@jlefebure Unless you're asking me to, and then yes I'd be happy to :P

**Comment by jlefebure** [comment]
Formsy is an awesome library. I'll be glad to contribute.

**Comment by rkuykendall** [comment]
@jlefebure Done! Let me know if you'd like another alpha release before your parameterized PR.

**Comment by rkuykendall** [comment]
@jlefebure Any time estimates on the parameterized value type PR? And do you think we should hold up the 2.x for it?

**Comment by jlefebure** [comment]
Hi, 
Yeah sorry. it's been a tough week. this feature should be included in the first release of Formsy V2, as many users (like me) are using  parameterized types with the Formsy mixin with TypeScript.  I'm starting working on that right now.

**Comment by jlefebure** [comment]
Ok, it seems to be working pretty. But I had to made several edits :
* PassDownProps is now built with WrappedProps and InjectedProps. InjectedProps are used to Omit its properties from component declaration
* Type Value has been removed, replaced by the parameterized value type (V)
* Validation function has been updated to do some type checking when necessary, and specify the type when necessary (example : isEmptyString only apply to string)
* The validations prop is now assignable to object, which allow more flexibility for validations rules specifications.
* For an unknown reason, the build is not including some utils functions. I had to remove the this pointer from inner methods calls in utils file

The withFormsy mixin now accept two type parameter : The component props, and the value type that the component will send and get via setValue and getValue. 

For a basic input text : 
```javascript
export default withFormsy<Props, string>(FormTextInput);
```
That work with every type, event if its not a string
```javascript
export default withFormsy<Props, Date | undefined>(FormDateInput);
```

This also make possible to use objects type. Our usecase is to specify an address object in an autocomplete input.
```javascript
export interface Address {
    identityComplement?: string;
    postalBox?: string;
    city?: string;
    zipcode?: string;
    countryCode?: string;
    geoComplement?: string;
    address1?: string;
    address2?: string;
    country?: string;
}

export default withFormsy<Props, Address>(AdvancedAddressInput);
```

Tests that I ran on our application seems to be ok, while i am not covering all Formsy usecases. I will submit a PR. Then, if it's ok for you, we can deploy a new alpha release, and I will schedule a test session with our non regression team to see if everything is working from our side.

**Comment by rkuykendall** [comment]
New type improvements in beta.8, which I'm considering a release client for 2.0. Unless something needs to be changed, I want to release.

**Comment by rkuykendall** [comment]
This has been fixed in the new 2.0.0 release. Thank you all for your patience! :tada: 

Please re-open this ticket or create a new one if I am mistaken.

# v2.0.0

- Much, much faster
- Many, many bug fixes
- Full Typescript typing
- Fixed deprecated lifecycle methods 
- Updated all libraries
- Fixed a number of legacy decisions in the Formsy API, mostly a reliance on function props over value props passed down to wrapped components. However, the API changes are minor and listed below:
  - `getErrorMessage()` => `errorMessage`
  - `getErrorMessages()` => `errorMessages`
  - `getValue()` => `value`
   `hasValue()` => `hasValue`
  - `isFormDisabled():` => `isFormDisabled`
  - `isFormSubmitted()` => `isFormSubmitted`
  - `isPristine()` => `isPristine`
  - `isRequired()` => `isRequired`
  - `isValid():` => `isValid`
  - `showError()` => `showError`
  - `showRequired()` => `showRequired`

https://github.com/formsy/formsy-react/compare/v1.1.4...v2.0.0
